Update trainer_utils.py

mzamini92 opened the pull request:
This modified version of the function includes a check for whether the output of function has a learning rate scheduler that needs to be updated based on the current batch size. If so, it updates the `num_batches` attribute of the scheduler to ensure that the learning rate is adjusted correctly.

# What does this PR do?
it can be one solution for the lr_scheduler not updated when auto_find_batch_size set to True and batch_size decays #21521 problem.


Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [+] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer: @stas00, Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

mzamini92 committed:
Update trainer_utils.py

This modified version of the function includes a check for whether the output of function has a learning rate scheduler that needs to be updated based on the current batch size. If so, it updates the `num_batches` attribute of the scheduler to ensure that the learning rate is adjusted correctly.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_23038). All of your documentation changes will be reflected on that endpoint.

sgugger commented:
cc @muellerzr

sgugger commented:
@muellerzr Could you have a look here?

muellerzr reviewed (approved):
This makes sense to me, @sgugger we might want to do this also in Accelerate, passing through `accelerator._schedulers` as a bool

muellerzr commented:
@mzamini92 could you rebase so we can double check no tests are breaking with this and we can merge? Thanks!

sgugger commented on src/transformers/trainer_utils.py:
```diff
@@ -653,7 +653,21 @@ def find_executable_batch_size(
+        except (RuntimeError, OverflowError, ValueError):
+            # Cut the batch size in half and try again
+            batch_size //= 2
+
```
Don't we need to break here if `batch_size` becomes 0?

mzamini92 replied:
@sgugger You are right. revised it.

muellerzr replied:
Actually, isn't this whole if else block redundant? And we're reducing the bs by 2^2 here, Accelerate needs to handle this. This PR should just touch the learning rate scheduler: https://github.com/huggingface/accelerate/blob/main/src/accelerate/utils/memory.py#L128-L142

mzamini92 replied:
Not sure if the mentioned snippet would actually cover everything but I guess in that case, if Accelerate should handle this then we can close this PR.

muellerzr replied:
Yes this is something Accelerate should handle, as the find_executable_batch_size function only should work if `auto_find_batch_size=True`. This was a good start, and the right idea, but it needs to be over there I'm afraid.

muellerzr replied:
(This wouldn't have been the case without the fact that the Trainer in torch runs on Accelerate now)

mzamini92 replied:
@muellerzr Please take look at this PR [#1632](https://github.com/huggingface/accelerate/pull/1632) and see if it makes sense for you.

mzamini92 replied:
@muellerzr I close this PR and take look at [#1633](https://github.com/huggingface/accelerate/pull/1633) PR. if it makes sense. if not you can close that one too. Thank you!

mzamini92 committed:
Update trainer_utils.py

Yes, it seems that there is a missing break statement in the code. In the `except` block, the `batch_size` is divided by 2 if any of the specified exceptions is raised. However, there is no break statement after that, so the execution will continue to the next iteration of the loop. If `batch_size` becomes 0, it will keep halving indefinitely.

mzamini92 commented:
> @mzamini92 could you rebase so we can double check no tests are breaking with this and we can merge? Thanks!

@muellerzr Thanks for reaching me. I did it based on Sylvain suggestion. please double check and I will revise if needed.
